### ISSUE TITLE
Bug: No reconciling Accessible Addons  of form spec variables

### ISSUE THREAD

**Issue by toschneck** [kubermatic/kubermatic#5993]
**User Story**

After a user change a parameter in the UI, the change get applied at the addon object in the cluster namespace, but the change of the parameter   
didn't get applied to the real add-on in the user cluster. The addon manifest of the user cluster won't get updated.

**At `https://run-2.lab.kubermatic.io/`:**
1. Add ingress addon
2. Increase size to 5 trough the edit button
![image.png](https://images.zenhubusercontent.com/5b6454d031c5fe7463837d8b/318fc022-43b2-4e0f-863c-cfa1b4bbee08)
3. see `k get addon -n cluster-wchcmmgwn8 ingress-nginx -o yaml`:
```
apiVersion: kubermatic.k8s.io/v1
kind: Addon
metadata:
  creationTimestamp: "2020-10-02T13:46:36Z"
  finalizers:
  - cleanup-manifests
  generation: 3
  labels:
    addons.kubermatic.io/ensure: "false"
  managedFields:
  - apiVersion: kubermatic.k8s.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers: {}
      f:status:
        f:conditions: {}
    manager: seed-controller-manager
    operation: Update
    time: "2020-10-02T13:46:39Z"
  - apiVersion: kubermatic.k8s.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .: {}
          f:addons.kubermatic.io/ensure: {}
        f:ownerReferences: {}
      f:spec:
        .: {}
        f:cluster:
          .: {}
          f:kind: {}
          f:name: {}
          f:uid: {}
        f:name: {}
        f:variables:
          .: {}
          f:replicas: {}
      f:status: {}
    manager: kubermatic-api
    operation: Update
    time: "2020-10-05T11:35:40Z"
  name: ingress-nginx
  namespace: cluster-wchcmmgwn8
  ownerReferences:
  - apiVersion: kubermatic.k8s.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: Cluster
    name: wchcmmgwn8
    uid: eebcda1d-7089-41ef-9097-8004eca82670
  resourceVersion: "178722254"
  selfLink: /apis/kubermatic.k8s.io/v1/namespaces/cluster-wchcmmgwn8/addons/ingress-nginx
  uid: 47e8c767-5788-4e89-9322-0de4c6b5fab7
spec:
  cluster:
    kind: Cluster
    name: wchcmmgwn8
    uid: eebcda1d-7089-41ef-9097-8004eca82670
  name: ingress-nginx
  variables:
    replicas: "5"
status:
  conditions:
  - lastHeartbeatTime: "2020-10-02T13:46:39Z"
    lastTransitionTime: "2020-10-02T13:46:39Z"
    status: "True"
    type: AddonResourcesCreatedSuccessfully

```
4. In user cluster manifest didn't get updated: ` k get deployments.apps -n ingress-nginx ingress-nginx-controller -o yaml `
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"controller","app.kubernetes.io/instance":"ingress-nginx","app.kubernetes.io/managed-by":"Helm","app.kubernetes.io/name":"ingress-nginx","app.kubernetes.io/version":"0.31.1","helm.sh/chart":"ingress-nginx-2.0.1","kubermatic-addon":"ingress-nginx"},"name":"ingress-nginx-controller","namespace":"ingress-nginx"},"spec":{"minReadySeconds":0,"replicas":1,"revisionHistoryLimit":10,"selector":{"matchLabels":{"app.kubernetes.io/component":"controller","app.kubernetes.io/instance":"ingress-nginx","app.kubernetes.io/name":"ingress-nginx"}},"template":{"metadata":{"labels":{"app.kubernetes.io/component":"controller","app.kubernetes.io/instance":"ingress-nginx","app.kubernetes.io/name":"ingress-nginx"}},"spec":{"containers":[{"args":["/nginx-ingress-controller","--publish-service=ingress-nginx/ingress-nginx-controller","--election-id=ingress-controller-leader","--ingress-class=nginx","--configmap=ingress-nginx/ingress-nginx-controller","--validating-webhook=:8443","--validating-webhook-certificate=/usr/local/certificates/cert","--validating-webhook-key=/usr/local/certificates/key"],"env":[{"name":"POD_NAME","valueFrom":{"fieldRef":{"fieldPath":"metadata.name"}}},{"name":"POD_NAMESPACE","valueFrom":{"fieldRef":{"fieldPath":"metadata.namespace"}}}],"image":"quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.31.1","imagePullPolicy":"IfNotPresent","lifecycle":{"preStop":{"exec":{"command":["/wait-shutdown"]}}},"livenessProbe":{"failureThreshold":3,"httpGet":{"path":"/healthz","port":10254,"scheme":"HTTP"},"initialDelaySeconds":10,"periodSeconds":10,"successThreshold":1,"timeoutSeconds":1},"name":"controller","ports":[{"containerPort":80,"name":"http","protocol":"TCP"},{"containerPort":443,"name":"https","protocol":"TCP"},{"containerPort":8443,"name":"webhook","protocol":"TCP"}],"readinessProbe":{"failureThreshold":3,"httpGet":{"path":"/healthz","port":10254,"scheme":"HTTP"},"initialDelaySeconds":10,"periodSeconds":10,"successThreshold":1,"timeoutSeconds":1},"resources":{"requests":{"cpu":"100m","memory":"90Mi"}},"securityContext":{"allowPrivilegeEscalation":true,"capabilities":{"add":["NET_BIND_SERVICE"],"drop":["ALL"]},"runAsUser":101},"volumeMounts":[{"mountPath":"/usr/local/certificates/","name":"webhook-cert","readOnly":true}]}],"dnsPolicy":"ClusterFirst","serviceAccountName":"ingress-nginx","terminationGracePeriodSeconds":300,"volumes":[{"name":"webhook-cert","secret":{"secretName":"ingress-nginx-admission"}}]}}}}
  creationTimestamp: "2020-10-02T13:46:37Z"
  generation: 1
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: ingress-nginx
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: ingress-nginx
    app.kubernetes.io/version: 0.31.1
    helm.sh/chart: ingress-nginx-2.0.1
    kubermatic-addon: ingress-nginx
  managedFields:
  - apiVersion: apps/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:labels:
          .: {}
          f:app.kubernetes.io/component: {}
          f:app.kubernetes.io/instance: {}
          f:app.kubernetes.io/managed-by: {}
          f:app.kubernetes.io/name: {}
          f:app.kubernetes.io/version: {}
          f:helm.sh/chart: {}
          f:kubermatic-addon: {}
      f:spec:
        f:progressDeadlineSeconds: {}
        f:replicas: {}
        f:revisionHistoryLimit: {}
        f:selector:
          f:matchLabels:
            .: {}
            f:app.kubernetes.io/component: {}
            f:app.kubernetes.io/instance: {}
            f:app.kubernetes.io/name: {}
        f:strategy:
          f:rollingUpdate:
            .: {}
            f:maxSurge: {}
            f:maxUnavailable: {}
          f:type: {}
        f:template:
          f:metadata:
            f:labels:
              .: {}
              f:app.kubernetes.io/component: {}
              f:app.kubernetes.io/instance: {}
              f:app.kubernetes.io/name: {}
          f:spec:
            f:containers:
              k:{"name":"controller"}:
                .: {}
                f:args: {}
                f:env:
                  .: {}
                  k:{"name":"POD_NAME"}:
                    .: {}
                    f:name: {}
                    f:valueFrom:
                      .: {}
                      f:fieldRef:
                        .: {}
                        f:apiVersion: {}
                        f:fieldPath: {}
                  k:{"name":"POD_NAMESPACE"}:
                    .: {}
                    f:name: {}
                    f:valueFrom:
                      .: {}
                      f:fieldRef:
                        .: {}
                        f:apiVersion: {}
                        f:fieldPath: {}
                f:image: {}
                f:imagePullPolicy: {}
                f:lifecycle:
                  .: {}
                  f:preStop:
                    .: {}
                    f:exec:
                      .: {}
                      f:command: {}
                f:livenessProbe:
                  .: {}
                  f:failureThreshold: {}
                  f:httpGet:
                    .: {}
                    f:path: {}
                    f:port: {}
                    f:scheme: {}
                  f:initialDelaySeconds: {}
                  f:periodSeconds: {}
                  f:successThreshold: {}
                  f:timeoutSeconds: {}
                f:name: {}
                f:ports:
                  .: {}
                  k:{"containerPort":80,"protocol":"TCP"}:
                    .: {}
                    f:containerPort: {}
                    f:name: {}
                    f:protocol: {}
                  k:{"containerPort":443,"protocol":"TCP"}:
                    .: {}
                    f:containerPort: {}
                    f:name: {}
                    f:protocol: {}
                  k:{"containerPort":8443,"protocol":"TCP"}:
                    .: {}
                    f:containerPort: {}
                    f:name: {}
                    f:protocol: {}
                f:readinessProbe:
                  .: {}
                  f:failureThreshold: {}
                  f:httpGet:
                    .: {}
                    f:path: {}
                    f:port: {}
                    f:scheme: {}
                  f:initialDelaySeconds: {}
                  f:periodSeconds: {}
                  f:successThreshold: {}
                  f:timeoutSeconds: {}
                f:resources:
                  .: {}
                  f:requests:
                    .: {}
                    f:cpu: {}
                    f:memory: {}
                f:securityContext:
                  .: {}
                  f:allowPrivilegeEscalation: {}
                  f:capabilities:
                    .: {}
                    f:add: {}
                    f:drop: {}
                  f:runAsUser: {}
                f:terminationMessagePath: {}
                f:terminationMessagePolicy: {}
                f:volumeMounts:
                  .: {}
                  k:{"mountPath":"/usr/local/certificates/"}:
                    .: {}
                    f:mountPath: {}
                    f:name: {}
                    f:readOnly: {}
            f:dnsPolicy: {}
            f:restartPolicy: {}
            f:schedulerName: {}
            f:securityContext: {}
            f:serviceAccount: {}
            f:serviceAccountName: {}
            f:terminationGracePeriodSeconds: {}
            f:volumes:
              .: {}
              k:{"name":"webhook-cert"}:
                .: {}
                f:name: {}
                f:secret:
                  .: {}
                  f:defaultMode: {}
                  f:secretName: {}
    manager: kubectl
    operation: Update
    time: "2020-10-02T13:46:37Z"
  - apiVersion: apps/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:deployment.kubernetes.io/revision: {}
      f:status:
        f:availableReplicas: {}
        f:conditions:
          .: {}
          k:{"type":"Available"}:
            .: {}
            f:lastTransitionTime: {}
            f:lastUpdateTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
          k:{"type":"Progressing"}:
            .: {}
            f:lastTransitionTime: {}
            f:lastUpdateTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
        f:observedGeneration: {}
        f:readyReplicas: {}
        f:replicas: {}
        f:updatedReplicas: {}
    manager: kube-controller-manager
    operation: Update
    time: "2020-10-02T13:47:07Z"
  name: ingress-nginx-controller
  namespace: ingress-nginx
  resourceVersion: "3104158"
  selfLink: /apis/apps/v1/namespaces/ingress-nginx/deployments/ingress-nginx-controller
  uid: d5c62ffa-a7a6-4f8d-b07e-02ce9101abcd
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/component: controller
      app.kubernetes.io/instance: ingress-nginx
      app.kubernetes.io/name: ingress-nginx
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: controller
        app.kubernetes.io/instance: ingress-nginx
        app.kubernetes.io/name: ingress-nginx
    spec:
      containers:
      - args:
        - /nginx-ingress-controller
        - --publish-service=ingress-nginx/ingress-nginx-controller
        - --election-id=ingress-controller-leader
        - --ingress-class=nginx
        - --configmap=ingress-nginx/ingress-nginx-controller
        - --validating-webhook=:8443
        - --validating-webhook-certificate=/usr/local/certificates/cert
        - --validating-webhook-key=/usr/local/certificates/key
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        image: quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.31.1
        imagePullPolicy: IfNotPresent
        lifecycle:
          preStop:
            exec:
              command:
              - /wait-shutdown
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 10254
            scheme: HTTP
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: controller
        ports:
        - containerPort: 80
          name: http
          protocol: TCP
        - containerPort: 443
          name: https
          protocol: TCP
        - containerPort: 8443
          name: webhook
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 10254
            scheme: HTTP
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          requests:
            cpu: 100m
            memory: 90Mi
        securityContext:
          allowPrivilegeEscalation: true
          capabilities:
            add:
            - NET_BIND_SERVICE
            drop:
            - ALL
          runAsUser: 101
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /usr/local/certificates/
          name: webhook-cert
          readOnly: true
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: ingress-nginx
      serviceAccountName: ingress-nginx
      terminationGracePeriodSeconds: 300
      volumes:
      - name: webhook-cert
        secret:
          defaultMode: 420
          secretName: ingress-nginx-admission
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-10-02T13:47:07Z"
    lastUpdateTime: "2020-10-02T13:47:07Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2020-10-02T13:46:37Z"
    lastUpdateTime: "2020-10-02T13:47:07Z"
    message: ReplicaSet "ingress-nginx-controller-7c8db695f8" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1

```

**Acceptance criteria**
- A change of the Addon parmater trigger an apply of the new rendered manifest in the user cluster (no matter if "enforce reconciling is selected or not)

**Comment by toschneck** [comment]
@zreigz could we may fix this also before the final 2.15 release?

**Comment by zreigz** [comment]
> @zreigz could we may fix this also before the final 2.15 release?

@toschneck right now addons use static files for the whole deployment. The variables are only stored in the CRD addon. To make it work we should use some YAML template for the deployments and apply variables to it. We have a discussion Oct 15 about this

**Comment by toschneck** [comment]
@zreigz we used the variable and a YAML template: https://github.com/kubermatic/customer-projects/blob/master/kubermatic-addons/custom-addon/ingress-nginx/deploy.yaml#L324
(I only copied the rendered yaml from the user cluster)

The problem is, if the stored value `replicas` in CRD is changed, the template don't get rendered and applied again. What leads to, that the user input of the UI don't get applied to the user cluster.

**Comment by maciaszczykm** [comment]
As @zreigz already mentioned it seems like the variables that user provides in the UI are only saved in the Addon CRD but not propagated to the user cluster. The mapping is missing too. Within @kubermatic/api-ui we had this initial idea that we can add `jsonPath` to the AddonConfig CRD's `formSpec` that will be a mapping between user-provided variable and the variable in the user cluster deployment.

**Comment by xrstf** [comment]
So I tested this again on the dev cluster.

- I created a new cluster.
- I added the node-exporter addon (for which I defined a new `Port` field, just so that I could have something to toggle). NMote that I enabled the `Continue reconciling this addon` option. This new `Port` field is not part of the templates, so changing it will have no effect on the rendered manifests (again, this is just for testing).

This then deploys the node-exporter into my cluster.

I now added a label to my DaemonSet and modified the Port count in the KKP dashboard. The seed-ctrlmgr logs

> Finished executing command        {"worker-name": "groucho", "cluster": "tv5xh5n8bq", "addon": "node-exporter", "cmd": "kubectl --kubeconfig /tmp/cluster-tv5xh5n8bq-addon-node-exporter-kubeconfig apply --prune -f /tmp/cluster-tv5xh5n8bq-node-exporter.yaml -l kubermatic-addon=node-exporter", "output": "daemonset.apps/node-exporter configured\nserviceaccount/node-exporter unchanged\nclusterrole.rbac.authorization.k8s.io/node-exporter unchanged\nclusterrolebinding.rbac.authorization.k8s.io/node-exporter unchanged\nservice/node-exporter unchanged\n"}

So the label survived the reconciling. If I however change an existing label (like the version number) or drop an existing label, they will get restored with the next reconciling (i.e. when something in the Addon changes or the 12h or so timeout is over).

TL;DR: Cannot reproduce the problem [anymore]. Addons are reconciled, but kubectl/Kubernetes merge stuff and so not all changes are overwritten by kubectl.
